### PR TITLE
[Issue #42] Allow a 'redirect_url' parameter for client registrations

### PR DESCRIPTION
**Pull request by ferjm**

### CODE REVIEW

File: doc/API.md [hunk 1]
```diff
@@ -110,14 +111,16 @@ Creates a new API client.
...
 
 ```ssh
 POST /api/clients HTTP/1.1
 Content-Type: application/json
 Cache-Control: no-cache
 Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImFkbWluIiwic2NvcGUiOiJhZG1pbiIsImlhdCI6MTQ3NDY0Nzc1OX0.R1vQOLVg8A-6i5QaZQVOGAzImiPvgAdkWiODYhYiNn4
 {
-	"name": "SensorWebClient"
+	"name": "SensorWebClient",
+  "redirectUrl": "https://domain.org/auth/complete"
```
Comment (julienw) [inline doc/API.md]: nit: looks like an indentation mismatch; I don't have any preference with/without, up to you :)

File: src/errors.js [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
-  ERRNO_FORBIDDEN               : 403,
-  ERRNO_RESOURCE_NOT_FOUND      : 404,
-  ERRNO_INTERNAL_ERROR          : 500
+  ERRNO_INVALID_API_CLIENT_NAME         : 100,
+  ERRNO_INVALID_API_CLIENT_REDIRECT_URL : 101,
+  ERRNO_BAD_REQUEST                     : 400,
+  ERRNO_UNAUTHORIZED                    : 401,
+  ERRNO_FORBIDDEN                       : 403,
+  ERRNO_RESOURCE_NOT_FOUND              : 404,
+  ERRNO_INTERNAL_ERROR                  : 500
```
Comment (julienw) [inline src/errors.js]: not for this patch, but I'm not a big fan of this indentation mechanism. As a result we need to change all lines when we really need to add 1 line...

File: src/models/clients.js [hunk 3]
```diff
@@ -22,6 +22,9 @@ module.exports = (sequelize, DataTypes) => {
       defaultValue: () => {
         return randomBytes(64).toString('hex');
       }
+    },
+    redirectUrl: {
```
Comment (julienw) [inline src/models/clients.js]: nit: tell me what you think, I would use "authRedirectUrl" to make it clear, when looking at the database, that this is for the authentication process.

File: src/routes/clients.js [hunk 4]
```diff
@@ -31,13 +33,26 @@ let router = express.Router();
...
+    let errno = ERRNO_BAD_REQUEST;
+    switch (error.param) {
+      case 'redirectUrl':
+        errno = ERRNO_INVALID_API_CLIENT_REDIRECT_URL;
+        break;
+      case 'name':
+        errno = ERRNO_INVALID_API_CLIENT_NAME;
+        break;
+    }
+    return ApiError(res, 400, errno, BAD_REQUEST);
```
Comment (julienw) [inline src/routes/clients.js]: (not for this patch either)

I don't really like how the errors are generated currently. For example I don't get why we need to put the numeric error number when errors.js already have it... It's also difficult to remember the order of the parameters.

Here is how I did it in a previous app:

* in the main app configuration file: https://github.com/project-abigail/calendar-server/blob/master/app.js#L46-L68
* here are my errors: https://github.com/project-abigail/calendar-server/blob/master/utils/errors.js
* here is an example about how to use them: https://github.com/project-abigail/calendar-server/blob/master/routes/users.js#L38-L41

File: src/routes/clients.js [hunk 5]
```diff
@@ -31,13 +33,26 @@ let router = express.Router();
...
 router.post('/', (req, res) => {
   // Required params: name
-  // XXX We will probably require redirect url for signin/signup flow.
   req.checkBody('name', 'missing or invalid required "name" parameter')
      .notEmpty();
 
-  const errors = req.validationErrors();
-  if (errors) {
-    return ApiError(res, 400, ERRNO_INVALID_API_CLIENT_NAME, BAD_REQUEST);
+  if (req.body.redirectUrl) {
```
Comment (julienw) [inline src/routes/clients.js]: same remark about the naming: `authRedirectUrl` ?

File: src/routes/clients.js [hunk 6]
```diff
@@ -31,13 +33,26 @@ let router = express.Router();
...
+  const error = req.validationErrors()[0];
+  if (error) {
+    let errno = ERRNO_BAD_REQUEST;
+    switch (error.param) {
+      case 'redirectUrl':
+        errno = ERRNO_INVALID_API_CLIENT_REDIRECT_URL;
+        break;
+      case 'name':
+        errno = ERRNO_INVALID_API_CLIENT_NAME;
+        break;
```
Comment (julienw) [inline src/routes/clients.js]: nit: I'd find it more readable to use a `default:` close to set `ERRNO_BAD_REQUEST`. (+ eslint will be happier when I'll set it up :) )

File: src/models/clients.js [hunk 7]
```diff
@@ -22,6 +22,9 @@ module.exports = (sequelize, DataTypes) => {
...
+      type: DataTypes.STRING(2048)
```
Comment (julienw) [inline src/models/clients.js]: `allowNull: false` ? Or a null means we can't login ? :)

Another question: do you think a client would want to configure several redirect urls ? (Facebook permits it)

File: src/routes/clients.js (hunk 5 above)
Comment (julienw) [inline src/routes/clients.js]: So it's optional ?

File: src/models/clients.js (hunk 3 above)
Comment (julienw) [inline src/models/clients.js]: I realize we also need a authFailureRedirectUrl (that could be null)

Comment (ferjm) [inline src/models/clients.js]: Sounds good. I'll change the naming and add the failure redirect url

File: src/errors.js (hunk 2 above)
Comment (ferjm) [inline src/errors.js]: Ok, fair enough

File: src/models/clients.js (hunk 7 above)
Comment (ferjm) [inline src/models/clients.js]: `allowNull: false` makes the field mandatory. I was thinking that we should probably allow creating new clients without auth redirection urls. Some of these clients might not want to use the APIs that will require user authentication (I think for now we will have only an API to bookmark sensors requiring user authentication). Some clients might only be interested on pushing sensors data to the server. And for sensors authentication we can't use a redirection url (we can't redirect to a sensor) so the authentication process will be different.

Allowing to configure several redirection urls sounds good.

File: src/routes/clients.js (hunk 5 above)
Comment (ferjm) [inline src/routes/clients.js]: Yes, as mentioned above, I think it should be optional.

File: src/routes/clients.js (hunk 6 above)
Comment (ferjm) [inline src/routes/clients.js]: Ok

File: src/routes/clients.js (hunk 4 above)
Comment (ferjm) [inline src/routes/clients.js]: I'll file an issue for this.

File: src/models/clients.js (hunk 7 above)
Comment (julienw) [inline src/models/clients.js]: ok for me !

File: src/routes/clients.js (hunk 4 above)
Comment (ferjm) [inline src/routes/clients.js]: #44

File: src/server.js [hunk 8]
```diff
@@ -18,7 +18,16 @@ import sensorthings     from './routes/sensorthings';
...
 
 app.use(bodyParser.urlencoded({ extended: false }));
-app.use(expressValidator());
+app.use(expressValidator({
+  customValidators: {
+    isArrayOfUrls: (value, options) => {
+      const validator = expressValidator.validator;
+      return Array.isArray(value) &&
+             value.length > 0 &&
+             value.every(item => validator.isUrl(item, options));
```
Comment (julienw) [inline src/server.js]: I realize that it is isURL we want here. I don't know why we didn't see it earlier, but now this fails badly for me.

File: test/test_clients_api.js [hunk 9]
```diff
@@ -75,11 +76,43 @@ describe('Clients API', () => {
...
+    });
+
+    it('should respond 201 Created if request is valid', done => {
       const name = 'clientName';
       server.post(endpointPrefix + '/clients')
             .set('Authorization', 'Bearer ' + token)
-            .send({ name: 'clientName' })
+            .send({ name: 'clientName',
+                    authRedirectUrls: ['http://something.com'],
+                    authFailureRedirectUrls: ['http://something.com'] })
```
Comment (julienw) [inline test/test_clients_api.js]: this should have found the issue; I don't know why this doesn't :/ (but locally it actually fails for me...)

Comment (julienw) [inline test/test_clients_api.js]: ah, this is interesting, this works with a clean node_modules. :)

Comment (ferjm) [inline test/test_clients_api.js]: Ugh... thanks for finding the issue.

**Pull request merged**